graphql_compiler/schema_transformation/utils.py

```diff
@@ -0,0 +1,135 @@
+    components, or if the schema contains some field of the query type that is named differently
+    from the type it queries.
+    """
+
+
+class SchemaNameConflictError(SchemaTransformError):
+    """Raised when renaming types or fields cause name conflicts."""
+
+
+def get_query_type_name(schema):
+    """Get the name of the query type of the input schema.
+
+    Args:
+        schema: GraphQLSchema
+
+    Returns:
+        str, name of the query type (e.g. RootSchemaQuery)
+    """
+    return schema.get_query_type().name
+
+
+def get_scalar_names(schema):
+    """Get names of all scalars used in the input schema.
+
+    Includes all user defined scalars, as well as any builtin scalars used in the schema; excludes
+    builtin scalars not used in the schema.
+
+    Returns:
+        Set[str], set of names of scalars used in the schema
+    """
+    type_map = schema.get_type_map()
+    scalars = {
+        type_name
+        for type_name in type_map
+        if isinstance(type_map[type_name], GraphQLScalarType)
+    }
+    return scalars
+
+
+class CheckQueryTypeFieldsNameMatchVisitor(Visitor):
+    """Check that every query type field's name is identical to the type it queries.
+
+    If not, raise SchemaStructureError.
+    """
+    def __init__(self, query_type):
+        """Create a visitor for checking query type field names.
+
+        Args:
+            query_type: str, name of the query type (e.g. RootSchemaQuery)
+        """
+        self.query_type = query_type
+        self.in_query_type = False
+
+    def enter_ObjectTypeDefinition(self, node, *args):
+        """If the node's name matches the query type, record that we entered the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = True
+
+    def leave_ObjectTypeDefinition(self, node, *args):
+        """If the node's name matches the query type, record that we left the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = False
+
+    def enter_FieldDefinition(self, node, *args):
+        """If inside the query type, check that the field and queried type names match.
+
+        Raises:
+            SchemaStructureError if the field name is not identical to the name of the type
+            that it queries
+        """
+        if self.in_query_type:
+            field_name = node.name.value
+            type_node = node.type
+            # NamedType node may be wrapped in several layers of NonNullType or ListType
+            while not isinstance(type_node, NamedType):
+                type_node = type_node.type
+            queried_type_name = type_node.name.value
+            if field_name != queried_type_name:
+                raise SchemaStructureError(
+                    u'Query type\'s field name "{}" does not match corresponding queried type '
+                    'name "{}"'.format(field_name, queried_type_name)
```

Comment:
another micro nit: you can add a `u` prefix before the single-quote here, and make the lines line up